File: backend/app/core/settings_manager.py
import json
import os
from typing import Dict, Any
import dotenv

SETTINGS_FILE = "system_settings.json"
ENV_FILE = ".env"

DEFAULT_SETTINGS = {
    "llm_model": "gemini-2.5-flash",
    "image_model": "gemini-2.5-flash-image",
    "comfyui": {
        "base_url": "http://127.0.0.1:8188",
        "enabled": False,
        "selected_workflow_file": None,
        "install_path": "D:\\ComfyUI"
    },
    "llm": {
        "provider": "gemini",
        "api_key": "",
        "base_url": "",
        "model": "gemini-2.5-flash"
    }
}

class SettingsManager:
    @classmethod
    def _get_file_path(cls) -> str:
        # Resolves to backend/system_settings.json assuming this file is in backend/app/core/
        base_dir = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
        return os.path.join(base_dir, SETTINGS_FILE)

    @classmethod
    def _get_env_path(cls) -> str:
        base_dir = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
        return os.path.join(base_dir, ENV_FILE)

    @classmethod
    def load_settings(cls) -> Dict[str, Any]:
        file_path = cls._get_file_path()
        env_path = cls._get_env_path()
        settings = DEFAULT_SETTINGS.copy()
        
        # 1. Load from file first
        if os.path.exists(file_path):
            try:
                with open(file_path, 'r', encoding='utf-8') as f:
                    file_settings = json.load(f)
                    # Deep merge for nested dictionaries
                    for key, value in file_settings.items():
                        if key in settings and isinstance(settings[key], dict) and isinstance(value, dict):
                            settings[key].update(value)
                        else:
                            settings[key] = value
            except (json.JSONDecodeError, IOError):
                pass

        # 2. Load from .env and apply overrides
        if os.path.exists(env_path):
            dotenv.load_dotenv(env_path)

        llm_provider_env = os.getenv("LLM_PROVIDER")
        if llm_provider_env:
            if "llm" not in settings:
                settings["llm"] = {}
            settings["llm"]["provider"] = llm_provider_env

        llm_api_key_env = os.getenv("LLM_API_KEY")
        if llm_api_key_env:
            if "llm" not in settings:
                settings["llm"] = {}
            settings["llm"]["api_key"] = llm_api_key_env
            
        llm_base_url_env = os.getenv("LLM_BASE_URL")
        if llm_base_url_env:
            if "llm" not in settings:
                settings["llm"] = {}
            settings["llm"]["base_url"] = llm_base_url_env

        llm_model_env = os.getenv("LLM_MODEL")
        if llm_model_env:
            if "llm" not in settings:
                settings["llm"] = {}
            settings["llm"]["model"] = llm_model_env

        return settings

    @classmethod
    def get(cls, key: str, default: Any = None) -> Any:
        settings = cls.load_settings()
        return settings.get(key, default)

    @classmethod
    def save_settings(cls, new_settings: Dict[str, Any]) -> Dict[str, Any]:
        current_settings = cls.load_settings()
        env_path = cls._get_env_path()
        
        # Ensure .env exists
        if not os.path.exists(env_path):
            open(env_path, 'a').close()
            
        # Extract secrets to .env
        if "llm" in new_settings:
            llm_settings = new_settings["llm"]
            
            # API Key
            if "api_key" in llm_settings:
                # Set in .env
                api_key = llm_settings["api_key"]
                if api_key: # Only save if not empty, or overwrite even if empty?
                    dotenv.set_key(env_path, "LLM_API_KEY", api_key)
                # Remove from new_settings so it doesn't go to json
                # Create a copy so we don't modify the input dict unexpectedly
                llm_settings_copy = llm_settings.copy()
                llm_settings_copy.pop("api_key", None)
                new_settings = new_settings.copy()
                new_settings["llm"] = llm_settings_copy

            # Optional: base_url to .env too if you want, but for now we do api_key
            
        # Now update current_settings with the safe new_settings
        # Need deep update so we don't overwrite the whole 'llm' dict
        for key, value in new_settings.items():
            if key in current_settings and isinstance(current_settings[key], dict) and isinstance(value, dict):
                current_settings[key].update(value)
            else:
                current_settings[key] = value

        # Remove api_key from current_settings if it's there before saving to JSON
        json_settings_to_save = current_settings.copy()
        if "llm" in json_settings_to_save and "api_key" in json_settings_to_save["llm"]:
            json_settings_to_save["llm"] = json_settings_to_save["llm"].copy()
            json_settings_to_save["llm"]["api_key"] = ""

        file_path = cls._get_file_path()
        try:
            with open(file_path, 'w', encoding='utf-8') as f:
                json.dump(json_settings_to_save, f, indent=4)
        except IOError as e:
            print(f"Error saving settings: {e}")
            pass
            
        # Return the merged settings including the secrets (which are now in current_settings or loaded from env)
        # Re-load to ensure we return exactly what's persisted/in-env
        return cls.load_settings()

File: backend/src/core/settings_manager.ts
import fs from 'fs';
import path from 'path';
import dotenv from 'dotenv';

const SETTINGS_FILE = 'system_settings.json';
const ENV_FILE = '.env';
const BACKEND_DIR = path.resolve(__dirname, '../../');

const DEFAULT_SETTINGS = {
    llm_model: 'gemini-2.5-flash',
    image_model: 'gemini-2.5-flash-image',
    comfyui: {
        base_url: 'http://127.0.0.1:8188',
        enabled: false,
        selected_workflow_file: null,
        install_path: 'D:\\ComfyUI'
    },
    advanced: {
        nsfw_enabled: false,
        pony_nsfw_lora: "Pony_Detail_Tweaker.safetensors",
        flux_nsfw_lora: "aidmaNSFWunlock.safetensors",
        nsfw_lora_strength: 0.8
    },
    llm: {
        provider: 'gemini',
        api_key: '',
        base_url: '',
        model: 'gemini-2.5-flash'
    }
};

export class SettingsManager {
    static getFilePath() {
        return path.join(BACKEND_DIR, SETTINGS_FILE);
    }

    static getEnvPath() {
        return path.join(BACKEND_DIR, ENV_FILE);
    }

    static loadSettings() {
        const filePath = SettingsManager.getFilePath();
        const envPath = SettingsManager.getEnvPath();
        const settings = JSON.parse(JSON.stringify(DEFAULT_SETTINGS)); // deep clone

        // 1. Load from file first
        if (fs.existsSync(filePath)) {
            try {
                const fileContent = fs.readFileSync(filePath, 'utf-8');
                const fileSettings = JSON.parse(fileContent);
                // Deep merge
                for (const [key, value] of Object.entries(fileSettings)) {
                    if (settings[key] && typeof settings[key] === 'object' && typeof value === 'object') {
                        settings[key] = { ...settings[key], ...value as Record<string, any> };
                    } else {
                        settings[key] = value;
                    }
                }
            } catch (err) {}
        }

        // 2. Load from .env and apply overrides
        if (fs.existsSync(envPath)) {
            dotenv.config({ path: envPath, override: true });
        }

        const llmProviderEnv = process.env.LLM_PROVIDER;
        if (llmProviderEnv) {
            settings.llm = settings.llm || {};
            settings.llm.provider = llmProviderEnv;
        }

        const llmApiKeyEnv = process.env.LLM_API_KEY;
        if (llmApiKeyEnv) {
            settings.llm = settings.llm || {};
            settings.llm.api_key = settings.llm.provider === 'ollama' ? 'ollama' : llmApiKeyEnv;
        }

        const llmBaseUrlEnv = process.env.LLM_BASE_URL;
        if (llmBaseUrlEnv) {
            settings.llm = settings.llm || {};
            settings.llm.base_url = llmBaseUrlEnv;
        }

        const llmModelEnv = process.env.LLM_MODEL;
        if (llmModelEnv) {
            settings.llm = settings.llm || {};
            settings.llm.model = llmModelEnv;
        }

        return settings;
    }

    static saveSettings(newSettings: Record<string, any>) {
        const currentSettings = SettingsManager.loadSettings();
        const envPath = SettingsManager.getEnvPath();

        if (!fs.existsSync(envPath)) {
            fs.writeFileSync(envPath, '');
        }

        let newSettingsCopy = JSON.parse(JSON.stringify(newSettings));
        let envContent = fs.readFileSync(envPath, 'utf-8');
        const originalEnvContent = envContent;
        const upsertEnvValue = (content: string, key: string, value: string) => {
            const pattern = new RegExp(`^${key}=.*$`, 'm');
            if (pattern.test(content)) {
                return content.replace(pattern, () => `${key}=${value}`);
            }
            const separator = content.length > 0 && !content.endsWith('\n') ? '\n' : '';
            return `${content}${separator}${key}=${value}\n`;
        };

        // Extract secrets to .env
        if (newSettingsCopy.llm && newSettingsCopy.llm.api_key !== undefined) {
            const apiKey = newSettingsCopy.llm.api_key;
            const isOllamaPlaceholder =
                newSettingsCopy.llm.provider === 'ollama' && apiKey === 'ollama';

            // Keep an existing cloud-provider secret when the local Ollama
            // placeholder is saved from the settings UI.
            if (!isOllamaPlaceholder) {
                envContent = upsertEnvValue(envContent, 'LLM_API_KEY', String(apiKey));
            }

            delete newSettingsCopy.llm.api_key;
        }

        // Keep environment overrides in sync with settings saved through the UI.
        // Otherwise stale LLM_* values would silently override system_settings.json.
        if (newSettingsCopy.llm) {
            const envMappings = [
                ['LLM_PROVIDER', newSettingsCopy.llm.provider],
                ['LLM_BASE_URL', newSettingsCopy.llm.base_url],
                ['LLM_MODEL', newSettingsCopy.llm.model]
            ] as const;

            for (const [key, value] of envMappings) {
                if (value !== undefined) {
                    envContent = upsertEnvValue(envContent, key, String(value));
                }
            }
        }

        if (envContent !== originalEnvContent) {
            fs.writeFileSync(envPath, envContent);
        }

        for (const [key, value] of Object.entries(newSettingsCopy)) {
            if (currentSettings[key] && typeof currentSettings[key] === 'object' && typeof value === 'object') {
                currentSettings[key] = { ...currentSettings[key], ...value as Record<string, any> };
            } else {
                currentSettings[key] = value;
            }
        }

        const jsonSettingsToSave = JSON.parse(JSON.stringify(currentSettings));
        if (jsonSettingsToSave.llm && jsonSettingsToSave.llm.api_key !== undefined) {
            jsonSettingsToSave.llm.api_key = '';
        }

        fs.writeFileSync(SettingsManager.getFilePath(), JSON.stringify(jsonSettingsToSave, null, 4), 'utf-8');
        return SettingsManager.loadSettings();
    }
}
